http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/response/CrtResponseAdapter.java

```diff
@@ -39,97 +40,110 @@ public final class CrtResponseAdapter implements HttpStreamResponseHandler {
-        return new CrtResponseAdapter(connection, responseComplete, request.sdkRequest(), request.readBufferSize());
+    private CrtResponseAdapter(HttpClientConnection connection,
+                               CompletableFuture<Void> completionFuture,
+                               SdkAsyncHttpResponseHandler responseHandler) {
+        this.connection = Validate.paramNotNull(connection, "connection");
+        this.completionFuture = Validate.paramNotNull(completionFuture, "completionFuture");
+        this.responseHandler = Validate.paramNotNull(responseHandler, "responseHandler");
     }
 
-    private void initRespBodyPublisherIfNeeded(HttpStream stream) {
-        if (respBodyPublisher == null) {
-            respBodyPublisher = new CrtResponseBodyPublisher(connection, stream, responseComplete, windowSize);
-        }
+    public static HttpStreamResponseHandler toCrtResponseHandler(HttpClientConnection crtConn,
+                                                                 CompletableFuture<Void> requestFuture,
+                                                                 SdkAsyncHttpResponseHandler responseHandler) {
+        return new CrtResponseAdapter(crtConn, requestFuture, responseHandler);
     }
 
     @Override
-    public void onResponseHeaders(HttpStream stream, int responseStatusCode, int blockType, HttpHeader[] nextHeaders) {
-        initRespBodyPublisherIfNeeded(stream);
-
-        for (HttpHeader h : nextHeaders) {
-            respBuilder.appendHeader(h.getName(), h.getValue());
+    public void onResponseHeaders(HttpStream stream, int responseStatusCode, int headerType, HttpHeader[] nextHeaders) {
+        if (headerType == HttpHeaderBlock.MAIN.getValue()) {
+            for (HttpHeader h : nextHeaders) {
+                responseBuilder.appendHeader(h.getName(), h.getValue());
+            }
         }
     }
 
     @Override
     public void onResponseHeadersDone(HttpStream stream, int headerType) {
         if (headerType == HttpHeaderBlock.MAIN.getValue()) {
-            initRespBodyPublisherIfNeeded(stream);
-
-            respBuilder.statusCode(stream.getResponseStatusCode());
-            sdkRequest.responseHandler().onHeaders(respBuilder.build());
-            sdkRequest.responseHandler().onStream(respBodyPublisher);
+            responseBuilder.statusCode(stream.getResponseStatusCode());
+            responseHandler.onHeaders(responseBuilder.build());
+            responseHandler.onStream(responsePublisher);
         }
     }
 
     @Override
     public int onResponseBody(HttpStream stream, byte[] bodyBytesIn) {
-        initRespBodyPublisherIfNeeded(stream);
+        CompletableFuture<Void> writeFuture = responsePublisher.write(ByteBuffer.wrap(bodyBytesIn));
+
+        if (writeFuture.isDone() && !writeFuture.isCompletedExceptionally()) {
+            // Optimization: If write succeeded immediately, return non-zero to avoid the extra call back into the CRT.
+            return bodyBytesIn.length;
```

Comment:
Yeah, I figured we could just use both mechanisms, so that when the consumer has enough demand to accept the request immediately we can avoid the extra JNI overhead of calling back into the CRT. Do you think that's confusing?

Reply:
Ah I see what you mean. That's smart.